### PR TITLE
refactor: use beacon chain deadlines for QBFT instance cleanup

### PR DESCRIPTION
**Pull request by diegomrsantos**
## Problem, Evidence, and Context (Required)

- `qbft_manager` was cleaning instances up on a fixed slot-based TTL instead of the actual duty validity window, which could remove live or already-decided instances before their beacon-chain deadline.
- For `SyncCommittee`, the previous same-slot cleanup was also tighter than the next-slot inclusion window that the client already models by starting consensus at `2/3` of the slot.
- This is worth doing now because it addresses the behavior reported in [#720](https://github.com/sigp/anchor/issues/720) and keeps the PR scoped to a bounded `qbft_manager` lifecycle improvement.
- Evidence: the previous retention model was based on an internal slot count rather than per-role beacon-chain inclusion windows.
- Relevant links: [#720](https://github.com/sigp/anchor/issues/720), [#914](https://github.com/sigp/anchor/issues/914) for the pre-existing late-caller race that remains out of scope here, and [#916](https://github.com/sigp/anchor/issues/916) for the broader timing-policy follow-up.

## Change Overview (Required)

- Replace fixed slot-based cleanup in `qbft_manager` with role-specific beacon-chain deadlines.
- Keep completed instances registered until their deadline instead of dropping them immediately.
- Keep sync-committee instances alive through the next slot rather than deleting them at the slot boundary.
- Add and tighten retention-focused tests so they prove the instance actually completes before asserting it remains registered.
- Add end-to-end sync-committee coverage using the real late-in-slot start offset and deadline cleanup path.
- Reviewer reading order: start with the deadline-based retention change, then the sync-committee deadline adjustment, then the retention tests.
- Intentionally did not change validator/message-validation timing policy outside `qbft_manager`, and did not try to solve the pre-existing late-caller race in [#914](https://github.com/sigp/anchor/issues/914). The broader timing-policy alignment work is tracked separately in [#916](https://github.com/sigp/anchor/issues/916).

## Risks, Trade-offs, and Mitigations (Required)

- Main risk is keeping QBFT instances alive longer than before and relying on deadline-based cleanup instead of the old fixed TTL.
- Trade-off: this improves `qbft_manager` retention semantics without yet centralizing all timing policy across other components.
- Risk is mitigated by keeping the behavior change scoped to `qbft_manager`, aligning `SyncCommittee` cleanup with the modeled next-slot inclusion window, and adding deadline-focused coverage, including an end-to-end sync-committee scenario.

## Validation (Required)

- `cargo test -p qbft_manager`
- `cargo fmt --all --check`
- `make lint`
- Commit hooks on the follow-up commits also ran:
  - `cargo fmt --all`
  - `cargo clippy --all`
  - `cargo sort workspace`

## Rollback (Required for behavior or runtime changes; optional otherwise)

- Safe to revert by reverting this PR's commits.
- No schema, config, or data migration impact.
- Revert would restore the old slot-based cleanup behavior.

## Blockers / Dependencies (Optional)

- N/A

## Additional Info / Next Steps (Optional)

- The late-caller race is tracked separately in [#914](https://github.com/sigp/anchor/issues/914).
- The broader timing-policy/source-of-truth follow-up is tracked in [#916](https://github.com/sigp/anchor/issues/916).
- The tester guidance now explicitly prefers `#[tokio::test(start_paused = true)]` when simulated time is safe, which matches the testing approach used in the new retention tests.

### CODE REVIEW

File: anchor/qbft_manager/src/lib.rs [hunk 1]
```diff
@@ -274,51 +331,84 @@ impl QbftManager {
...
+                Some(id) = completion_rx.recv() => {
+                    match id {
+                        InstanceId::BeaconVote(id) => {
+                            self.beacon_vote_instances.remove(&id);
+                        }
+                        InstanceId::ValidatorConsensus(id) => {
+                            self.validator_consensus_data_instances.remove(&id);
+                        }
+                    }
+                }
```
Comment (dknopik) [inline anchor/qbft_manager/src/lib.rs]: There is a problem with this approach:

In theory, there might be a race condition where some tasks try to register their oneshot channel to an instance after it has completed. This might e.g. be the case if multiple validator attestation duties wait for the same committee instance. If we start the instance late (e.g. because of a struggling BN), the first thread will start the instance, which might complete immediately due to replayed messages, giving no opportunity for the other tasks to register their listeners. This is why the current code cleans up at a fixed time regardless of completion.

Instead, we could move the cleanup time in this branch - to give some time (til end of next slot?) to get the instance result. Wdyt?

Comment (diegomrsantos) [reply anchor/qbft_manager/src/lib.rs]: I need more context to understand what's described in the first paragraph.

File: anchor/qbft_manager/src/tests.rs [hunk 2]
```diff
@@ -902,6 +902,55 @@ mod manager_tests {
...
+        // This will cause the instance to go through multiple round changes.
+        context.set_operators_offline(&[2, 3, 4]);
+
+        // Advance time and slots to simulate reaching round 10
+        // Instance starts at slot 0
+        let slot_duration = Duration::from_secs(12);
+
+        // Advance through multiple slots while QBFT progresses
+        // This triggers cleanup logic which should NOT remove the active instance
+        for slot in 1..=50 {
```
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply anchor/qbft_manager/src/tests.rs]: No longer relevant.

File: anchor/qbft_manager/src/tests.rs [hunk 3]
```diff
@@ -902,6 +902,55 @@ mod manager_tests {
...
+            clock.set_slot(slot);
+            tokio::time::sleep(slot_duration).await;
+
```
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply anchor/qbft_manager/src/tests.rs:1247]: No longer relevant — the comment math has been corrected in the current code.

File: anchor/qbft_manager/src/tests.rs [hunk 4]
```diff
@@ -902,6 +902,55 @@ mod manager_tests {
...
+    #[tokio::test(start_paused = true)]
+    // Test that Committee instances can reach late rounds (9+) with max_round=12 configuration.
+    // This verifies that instances survive long enough to progress through many round changes
+    // as configured. Committee role has max_round=12, so instances should be able to reach
+    // round 10 before timing out at round 13.
+    //
+    // The test simulates network conditions where consensus cannot be reached early by keeping
+    // all but one operator offline, forcing round changes. We advance the slot to trigger
+    // cleanup and verify the instance survives to reach round 10.
+    async fn test_committee_can_reach_late_rounds() {
```
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply anchor/qbft_manager/src/tests.rs:1222]: No longer relevant — the test now passes with the expiry-based cleanup; no `#[ignore]` needed.

File: anchor/qbft_manager/src/tests.rs [hunk 5]
```diff
@@ -902,6 +903,55 @@ mod manager_tests {
...
+        let slot_duration = Duration::from_secs(12);
+
+        // Advance through multiple slots while QBFT progresses
+        // This triggers cleanup logic which should NOT remove the active instance
+        for slot in 1..=50 {
+            clock.set_slot(slot);
+            tokio::time::sleep(slot_duration).await;
+
+            // At slot 22 (256 seconds = 16s + 240s), we should be around round 10
+            // Rounds 1-8: 16s, Rounds 9-10: 240s = 256s total
```
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply anchor/qbft_manager/src/tests.rs]: No longer relevant — duplicate of an earlier comment; math has been corrected.

File: anchor/qbft_manager/src/tests.rs [hunk 6]
```diff
@@ -902,6 +903,55 @@ mod manager_tests {
...
+        // This will cause the instance to go through multiple round changes.
+        context.set_operators_offline(&[2, 3, 4]);
+
+        // Advance time and slots to simulate reaching round 10
+        // Instance starts at slot 0
+        let slot_duration = Duration::from_secs(12);
+
+        // Advance through multiple slots while QBFT progresses
+        // This triggers cleanup logic which should NOT remove the active instance
+        for slot in 1..=50 {
```
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply anchor/qbft_manager/src/tests.rs]: No longer relevant — loop bound was already reduced to `1..=25`.

File: anchor/qbft_manager/src/lib.rs [hunk 7]
```diff
@@ -98,8 +115,20 @@ pub struct QbftInitialization<D: QbftData> {
     on_completed: oneshot::Sender<Completed<D>>,
 }
 
-// Map from an identifier to a sender for the instance
-type Map<I, D> = DashMap<I, UnboundedSender<QbftMessage<D>>>;
+// Manager's bookkeeping for an instance
+pub struct ManagedInstance<D: QbftData> {
```
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply anchor/qbft_manager/src/lib.rs:158]: No longer relevant — `ManagedInstance` was simplified to a single `sender` field with an adequate doc comment.

File: anchor/qbft_manager/src/lib.rs [hunk 8]
```diff
@@ -39,8 +39,25 @@ const QBFT_INSTANCE_NAME: &str = "qbft_instance";
 const QBFT_MESSAGE_NAME: &str = "qbft_message";
 const QBFT_CLEANER_NAME: &str = "qbft_cleaner";
 
-/// Number of slots to keep before the current slot
-const QBFT_RETAIN_SLOTS: u64 = 1;
+/// Calculate the beacon chain inclusion deadline for a duty
+fn calculate_deadline(role: Role, slot: types::Slot, slots_per_epoch: u64) -> types::Slot {
+    match role {
+        Role::Committee | Role::Aggregator => {
```
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply anchor/qbft_manager/src/lib.rs]: No longer relevant — `calculate_deadline` already has the enhanced EIP-7045 comment with the full calculation breakdown.

File: anchor/qbft_manager/src/tests.rs [hunk 9]
```diff
@@ -902,6 +902,55 @@ mod manager_tests {
...
+            setup.executor,
+            CommitteeSize::Four,
+            setup.all_data,
+        )
+        .await;
+
+        // Keep 3 operators offline initially to prevent consensus and force round changes.
+        // With only 1 operator online out of 4, we cannot reach quorum (need 3).
+        // This will cause the instance to go through multiple round changes.
+        context.set_operators_offline(&[2, 3, 4]);
```
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply anchor/qbft_manager/src/tests.rs:1236]: Addressed — added a named constant (`ROUND_11_SLOT`) with a documented timing calculation, and an intermediate assertion inside the loop that verifies the instance remains in the registry at every slot.

File: anchor/qbft_manager/src/lib.rs (hunk 1 above)
Comment (diegomrsantos) [reply anchor/qbft_manager/src/lib.rs]: Good catch. This race exists, but it is pre-existing, not introduced by this PR.

Before this PR, the instance task stays alive in `Decided` state and the registry entry lingers until the slot-based cleaner removes it. Once the cleaner drops the entry, the `tx` is dropped, `rx.recv()` returns `None`, and the task exits. A late caller after that point hits `Vacant`, spawns a new instance, and hangs until timeout. The grace window is larger under the old cleanup scheme, but the underlying late-caller behavior is still there.

This PR does make that behavior easier to hit by breaking out of the loop on `Decided` and removing the entry immediately via completion notification.

The broader issue is that independent code paths can call `decide_instance` for the same `CommitteeInstanceId` at different times. Grouping validators per committee, as in [#834](https://github.com/sigp/anchor/pull/834), improves that by reducing duplicate local callers within a duty path. Applying the same pattern to sync committee signing would help for the same reason.

But that change alone does not fully eliminate the late-caller behavior; that still depends on the cleanup / lifecycle semantics in `qbft_manager`.

Comment (dknopik) [reply anchor/qbft_manager/src/lib.rs]: We could modify `ManagedInstance` to either hold a channel sender to a running instance, or a resulting value `D` obtained from a finished instance to accommodate late callers. This would allow us to clean up the instance as soon as it is finished by dropping the sender and storing the finished value. I am unsure how long the resulting value should be kept - 1 slot?

As you said, the underlying root cause is the possibility of late callers, but I am unsure how to prevent this without a major refactor.

Comment (diegomrsantos) [reply anchor/qbft_manager/src/lib.rs]: Good suggestion. What we can do in this PR is remove the immediate cleanup on completion (Branch 1 in the cleaner) and the `break` after `Decided` in the instance task — so the instance stays alive until its beacon chain deadline expires. The `Decided` state already handles late `Initialize` messages by returning the cached result (see `instance.rs` line 79-83), so late callers would get the decided value immediately.

The pre-existing late-caller race (where callers arrive after the entry is removed entirely) is a separate concern — I'll create an issue to track that for a follow-up PR.

**Comment by diegomrsantos** [comment]
@claude do a final review

**Comment by diegomrsantos** [comment]
@claude It looks like the latest versions of both branches weren't used in your review. Do it again

**Comment by diegomrsantos** [comment]
Follow-up to Claude's note about the `SyncCommittee` same-slot deadline: this is now fixed on the PR branch.

- `0968f4950` changes `SyncCommittee` cleanup to keep instances alive through the next-slot inclusion window while leaving `Proposer` same-slot.
- `0968f4950` also adds an end-to-end paused-time test that uses the real late-in-slot start offset, crosses the first slot boundary without quorum, then proves the instance still reaches consensus and is cleaned at the following deadline boundary.
- `1035f58bd` fixes the resulting CI `clippy::clone_on_copy` failure in the new test helper.
- `2681708e8` rewrites the new sync-committee test into a clearer Arrange / Act / Assert structure.

Validation after the follow-ups:
- `make lint`
- `cargo test -p qbft_manager`
